src/Transform/ONNX/LayerNameToLocationPass.cpp

```diff
@@ -0,0 +1,83 @@
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Pass/Passes.hpp"
+
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/Pass/Pass.h"
+
+#include "llvm/ADT/StringRef.h"
+
+#include <memory>
+
+using namespace mlir;
+using namespace llvm;
+
+namespace onnx_mlir {
+namespace {
+struct LayerNameToLocationPass
+    : public PassWrapper<LayerNameToLocationPass, OperationPass<>> {
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(LayerNameToLocationPass)
+
+  LayerNameToLocationPass() = default;
+  LayerNameToLocationPass(const LayerNameToLocationPass &pass) = default;
+
+  [[nodiscard]] StringRef getArgument() const override {
+    return "onnx-layer-name-location";
+  }
+
+  [[nodiscard]] StringRef getDescription() const override {
+    return "Inserts the ONNX layer name of each operation into the location "
+           "info associated with it.";
+  }
+
+  void runOnOperation() final;
+};
+
+void LayerNameToLocationPass::runOnOperation() {
+  // Check whether op has a onnx_node_name attribute and put that on the
+  // existing location info.
+  Operation *op = getOperation();
+  MLIRContext *context = op->getContext();
+  Dialect *onnxDialect = context->getLoadedDialect("onnx");
+  if (!onnxDialect) {
+    op->emitError("failed to load ONNX dialect");
+    return signalPassFailure();
+  }
+
+  // Counter and names to make invalid locations unique
+  unsigned invLocSeq = 0;
+  StringRef invLocName = "INVALID:FXML-1477:";
+
+  op->walk([&](Operation *nestedOp) {
+    Location loc = nestedOp->getLoc();
+    NameLoc nameLoc;
+
+    // Extend the existing location with the layer name, if available
+    if (auto layerName =
+            nestedOp->getAttrOfType<StringAttr>("onnx_node_name")) {
+      nameLoc = NameLoc::get(layerName, loc);
+    }
+
+    // All onnx ops (except constants) are expected to have an onnx_node_name.
+    // If onnx_node_name is not available and this is an onnx operation, extend
+    // the location with a name to indicate a missing location.
+    else if (nestedOp->getDialect() == onnxDialect &&
+             !isa<ONNXConstantOp>(nestedOp)) {
```

Comment:
```suggestion
    else if (!isa<ONNXConstantOp>(nestedOp)) {
```
Is there any reason you first check the dialect and then the operation kind? Because ONNXConstantOp is part of that dialect already.